set-2 cases and enhancing set-1 smartswitch cases

nissampa opened the pull request:
==========================================================================================

Test Plan Link:
https://github.com/sonic-net/sonic-mgmt/blob/master/docs/testplan/Smartswitch-test-plan.md

Test Cases in this PR:
1.4 Check DPU console
1.8 Check the NTP date and timezone between DPU and NPU
1.9 Check the State of DPUs
1.10 Check the Health of DPUs
1.13 Check Memory on DPU
1.14 Check DPU status and pcie Link after memory exhaustion on Switch
1.15 Check DPU status and pcie Link after memory exhaustion on DPU
1.16 Check DPU status and pcie Link after restart pmon
1.18 Check DPU status and pcie Link after kernel panic on Switch
1.19 Check DPU status and pcie Link after kernel panic on DPU

==========================================================================================

File: smartswitch/platform_tests/test_platform_dpu.py
Changed the file name from test_show_platform_dpu.py to test_platform_dpu.py
NewTest Cases:

test_restart_pmon
test_system_health_state
test_dpu_console
test_npu_dpu_date
test_dpu_memory
test_system_health_summary

==========================================================================================

File: smartswitch/platform_tests/test_reload_dpu.py
Changed the name for existing test cases.
NewTest Cases: 

test_dpu_status_post_switch_mem_exhaustion
test_dpu_status_post_switch_kernel_panic
test_dpu_status_post_dpu_kernel_panic
test_dpu_check_post_dpu_mem_exhaustion

==========================================================================================

UT Logs:

[smartswitch_PR_logs.zip](https://github.com/user-attachments/files/18119654/smartswitch_PR_logs.zip)

==========================================================================================

mssonicbld commented:
/azp run

nissampa commented:
/azpw run

nissampa commented:
/azpw run

mssonicbld commented:
/AzurePipelines run

mssonicbld commented:
/azp run

vvolam commented on tests/smartswitch/common/device_utils_dpu.py:
```diff
@@ -132,35 +134,198 @@ def check_dpu_module_status(duthost, power_status, dpu_name):
             return True
         else:
-            logging.info("'{}' is online ...".format(dpu_name))
+            logging.info("'{}' is offline ...".format(dpu_name))
```
What is the difference in logs b/w if and else?

nissampa replied:
Basically it is a combination.

Input : dpu name and status (that we wanted it to be):
Output: True or False based on the combination.

Expected DPU State (Online/Offline) vs Actual DPU state (Online/Offline)

This was designed as per the comments given in previous PR. 
I noticed the log was not representing properly and hence the change here.

vvolam replied:
Yes, I totally agree with you. What I mean was log line at 134 and 137 is same, so instead of repeating the code, log could be moved to line 133 before "if power_status == "off":".

nissampa replied:
Oh yeah, I removed it.

vvolam commented on tests/smartswitch/common/device_utils_dpu.py:
```diff
@@ -132,35 +134,198 @@ def check_dpu_module_status(duthost, power_status, dpu_name):
             return True
         else:
-            logging.info("'{}' is offline ...".format(dpu_name))
+            logging.info("'{}' is online ...".format(dpu_name))
```
same comment as above.

nissampa replied:
Same reply.

nissampa replied:
done.

vvolam commented on tests/smartswitch/platform_tests/test_reload_dpu.py:
```diff
@@ -22,32 +21,42 @@
 
 def test_dpu_ping_after_reboot(duthosts, enum_rand_one_per_hwsku_hostname,
-                               localhost, platform_api_conn, num_dpu_modules):    # noqa F811
+                               localhost, platform_api_conn,  # noqa F811
```
F811 error could be removed by removing the import at line 16 per my understanding.

nissampa replied:
F811 was introduced here in this commit by yotongzhang-microsoft on smartswitch files as well. 
So, I followed the same suit.

https://github.com/sonic-net/sonic-mgmt/commit/475f52f47fe2e8ecb9d786d3194e1c04fbcc883c#diff-1ff5cd58b35cd3c24b8e95e46748a838458539db3f66b1bf0915980c82d94416

vvolam commented on tests/smartswitch/common/device_utils_dpu.py:
```diff
@@ -3,6 +3,7 @@
 import pytest
+import re
 from tests.common.devices.sonic import *  # noqa: F401,F403
 from tests.common.platform.device_utils import platform_api_conn  # noqa: F401,F403
```
This import is not required. You can also remove "noqa" tags where platform_api_conn is used. For example,

-def num_dpu_modules(platform_api_conn):   # noqa F811
+def num_dpu_modules(platform_api_conn):

nissampa replied:
F811 was introduced here in this commit by yotongzhang-microsoft on smartswitch files as well.
So, I followed the same suit.

https://github.com/sonic-net/sonic-mgmt/commit/475f52f47fe2e8ecb9d786d3194e1c04fbcc883c#diff-1ff5cd58b35cd3c24b8e95e46748a838458539db3f66b1bf0915980c82d94416

vvolam replied:
Okay, can you please fix it in the files you are modifying? As I mentioned previously in one of the comment, this can be simply avoided by removing import of platform_api_conn line at 7. Please give it a try. I have already tried it locally and that works.

nissampa replied:
Can we go ahead and remove line 7 totally ?

vvolam replied:
yes

nissampa replied:
I will change it, test it and update it here.

nissampa replied:
If I remove that line, it gives me error while running the test cases.
If I remove the F811, it gives me error on flake8

vvolam replied:
@nissampa I still think it is good to get rid of this as well. But leave it to your call.

vvolam commented on tests/smartswitch/platform_tests/test_show_platform_dpu.py:
```diff
@@ -4,6 +4,7 @@
+from datetime import datetime
 from tests.common.utilities import wait_until
 from tests.common.helpers.assertions import pytest_assert
 from tests.smartswitch.common.device_utils_dpu import *  # noqa: F403,F401,E501
```
Please import only used functions and resolve all flake8 errors. Also chassis import is done but not used. That can be removed.

Additionally no function is getting used from tests.common.devices.sonic, please remove it.

nissampa replied:
chassis and modules are imported to use all the apis related to that in the tests.

vvolam reviewed (commented):
Please cleanup all flake8 "noqa" tags.

vvolam commented on tests/smartswitch/platform_tests/test_show_platform_dpu.py:
```diff
@@ -117,13 +123,234 @@ def test_pcie_link(duthosts, enum_rand_one_per_hwsku_hostname,
+
+    logging.info("Checking pmon status")  # noqa: F405
+    pmon_status = check_pmon_status(duthost)  # noqa: F405
+    pytest_assert(pmon_status == 1, "PMON status is UP")
```
Should the failing message be "PMON container is DOWN?"

nissampa replied:
Changed it.

vvolam commented on tests/smartswitch/platform_tests/test_show_platform_dpu.py:
```diff
@@ -117,13 +123,234 @@ def test_pcie_link(duthosts, enum_rand_one_per_hwsku_hostname,
+                          module.get_midplane_ip(platform_api_conn, index))
+
+    ping_status = check_dpu_ping_status(duthost, ip_address_list)  # noqa: F405, E501
+    pytest_assert(ping_status == 1, "Ping to DPU has been tested")
```
Same here. Failure message seems to be wrong. In case of ping failure, log a message like "Ping to one or more DPUs failed". Please recheck other assertions logs in the PR.

nissampa replied:
Changed here and in all the respective places.

vvolam reviewed (commented):
Please recheck the assertion logs, most of them seems to be opposite in case of assertion is raised on failure.

nissampa commented on tests/smartswitch/platform_tests/test_reload_dpu.py:
```diff
@@ -22,32 +21,42 @@
 
```
Removed the config imports as suggested.

nissampa replied:
Regarding removal of noqa for imports device_utils_dpu, I started off using * for keep on adding more functions.
I thought it would be otherwise that importing everything would make it more messy.

vvolam replied:
@nissampa Importing everything deals with only couple of lines but adding "noqa" tags everytime a function is called will make it look messier in my opinion.

nissampa replied:
Agreed. I will change it.

nissampa commented on tests/smartswitch/platform_tests/test_reload_dpu.py:
```diff
@@ -22,32 +21,42 @@
+                                                 platform_api_conn)
+
+    logging.info("Checking DPU connectivity before reboot..")
+    pytest_assert(wait_until(30, 10, 0, check_dpu_ping_status,  # noqa: F405
```
Yes changing it.

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

vvolam commented on tests/smartswitch/platform_tests/test_reload_dpu.py:
```diff
@@ -10,14 +10,15 @@
      execute_dpu_commands,
      check_dpu_reboot_cause,
-     num_dpu_modules
+     num_dpu_modules  # noqa: F401
```
why do you need noqa here?

nissampa replied:
This place I removed it.

vvolam commented on tests/smartswitch/platform_tests/test_reload_dpu.py:
```diff
@@ -4,16 +4,22 @@
-from tests.smartswitch.common.device_utils_dpu import *  # noqa: F401,F403,E501
-from tests.common.helpers.platform_api import chassis, module  # noqa: F401
+from tests.common.helpers.platform_api import module
+from tests.smartswitch.common.device_utils_dpu import (  # noqa: F401
```
Do we still need to add noqa here? What is the error otherwise?

nissampa replied:
This was similar to that line 7 of importing platform_conn.

vvolam replied:
Didn't get you? What is the Flake8 error you are getting here?

nissampa replied:
Since importing fixtures as part of import, it gives me flake8 error. Without import test is giving me error.

vvolam commented on tests/smartswitch/platform_tests/test_reload_dpu.py:
```diff
@@ -22,53 +28,277 @@
+    logging.info("Checking for Interface status")
+    pytest_assert(wait_until(300, 5, 0, check_interface_status_of_up_ports,
+                  duthost),
+                  "Not all ports that are admin up on are operationally up")
```
nit: instead use "Not all ports that are admin up, are operationally up"

vvolam commented on tests/smartswitch/platform_tests/test_reload_dpu.py:
```diff
@@ -22,53 +28,277 @@
+    logging.info("Checking for Interface status")
+    pytest_assert(wait_until(300, 5, 0, check_interface_status_of_up_ports,
+                  duthost),
+                  "Not all ports that are admin up on are operationally up")
```
nit: instead use "Not all ports that are admin up, are operationally up"

nissampa replied:
Changed it.

vvolam commented on tests/smartswitch/platform_tests/test_reload_dpu.py:
```diff
@@ -22,53 +28,277 @@
+                  executable="/bin/bash")
+
+    logging.info("Waiting for ssh connection to switch")
+    wait_for_startup(duthost, localhost, 100, 400)
```
Is this delay and timeout in secs? Is this sufficient for all vendors? Also, could you please add a comment above about the time metric used here?

@oleksandrivantsiv - for Viz

nissampa replied:
All time delays are in secs.

vvolam commented on tests/smartswitch/platform_tests/test_reload_dpu.py:
```diff
@@ -22,53 +28,277 @@
+                             localhost,
+                             platform_api_conn, num_dpu_modules):  # noqa: F811, E501
+    """
+    @summary: To Verify `kernel panic on dpu`
```
nit: Please use something like "Test to verify DPU recovery on "kernel panic on DPU""?? for better readability

nissampa replied:
Changed it.

vvolam commented on tests/smartswitch/platform_tests/test_reload_dpu.py:
```diff
@@ -22,53 +28,277 @@
+    ip_address_list = []
+    kernel_panic_cmd = "echo c | sudo tee /proc/sysrq-trigger"
+
+    for index in range(num_dpu_modules):
```
Can we do parallel check on all DPUs here to save time? This will save time in test to wait for each DPU in serial.

nissampa replied:
We can do that. I will modify both the tests cases where it involves dpu.

vvolam commented on tests/smartswitch/platform_tests/test_reload_dpu.py:
```diff
@@ -22,53 +28,277 @@
+
+        logging.info("Shutting down %s" % (dpu_name))
+        duthosts.shell("config chassis modules shutdown %s" % (dpu_name))
+        pytest_assert(wait_until(360, 120, 0,
```
Is this time sufficient for all vendors? Can we define macros for all the time numeric values in the tests and add a description for easy modifications?

nissampa replied:
All the time metrics are all in secs. I used same time for interfaces and startups that are used in previous existing cases. Only new thing here in the test cases are for DPU up and down times.

vvolam replied:
Could you please define macros for these metrics for easy modification or the values? This helps in avoiding the mistake of not modifying at some places in future?

nissampa replied:
Yes noted. I used the values particularly adjusting for DPUs by running the test many a times and come up with that values.

vvolam commented on tests/smartswitch/platform_tests/test_reload_dpu.py:
```diff
@@ -22,53 +28,277 @@
+                                  localhost,
+                                  platform_api_conn, num_dpu_modules):  # noqa: F811, E501
+    """
+    @summary: To Verify `kernel panic on dpu`
```
Is this kernel panic or memory exhaustion on DPU?

nissampa replied:
It is memory exhaustion on DPU and I changed it.

vvolam commented on tests/smartswitch/platform_tests/test_reload_dpu.py:
```diff
@@ -22,53 +28,277 @@
+
+    logging.info("Checking all powered on DPUs connectivity")
+    ping_status = check_dpu_ping_status(duthost, ip_address_list)
+    pytest_assert(ping_status == 1, "Ping to DPU has failed")
```
nit: Please modify log message as "Ping to one or more DPUs has failed"?

nissampa replied:
Changed it.

vvolam commented on tests/smartswitch/platform_tests/test_show_platform_dpu.py:
```diff
@@ -100,30 +114,245 @@ def test_pcie_link(duthosts, enum_rand_one_per_hwsku_hostname,
+
+        command = ('python -c "import pexpect; '
+                   'child = pexpect.spawn(\'python /usr/local/bin/dpu-tty.py -n dpu%s\'); '  # noqa: E501
+                   'child.expect(r\' \'); '
```
Is the expectation same on all vendor devices?

@oleksandrivantsiv for viz

nissampa replied:
It should be same. I am just checking the prompt till the login for DPU.

gpunathilell replied:
It should work for Nvidia platforms as well

vvolam replied:
Thank you. I have verified myself and working on both platforms.

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

nissampa commented:
@vvolam, Let me know if changes are good.

nissampa commented:
@oleksandrivantsiv, Kindly please pass this PR to Nvidia test team.

oleksandrivantsiv commented:
@congh-nvidia, @JibinBao please review this PR

vvolam commented on tests/smartswitch/common/device_utils_dpu.py:
```diff
@@ -89,8 +95,10 @@ def dpu_power_on(duthost, platform_api_conn, num_dpu_modules):    # noqa F811
-    pytest_assert(wait_until(180, 60, 0, check_dpu_ping_status,  # noqa: F405
-                  duthost, ip_address_list), "Not all DPUs are operationally up")
+    pytest_assert(wait_until(PING_MAX_TIMEOUT, PING_TIME_INTV, 0,
+                  check_dpu_ping_status,  # noqa: F405
```
@nissampa can you get rid of this flake warning?

nissampa replied:
Removed it.

vvolam commented on tests/smartswitch/platform_tests/test_show_platform_dpu.py:
```diff
@@ -100,30 +119,245 @@ def test_pcie_link(duthosts, enum_rand_one_per_hwsku_hostname,
+    pytest_assert(pmon_status == 1, "PMON status is Not UP")
+
+
+def test_system_health_state(duthosts, enum_rand_one_per_hwsku_hostname,
```
@nissampa could you split it into small functions for readability

nissampa replied:
Changed it.

KrisNey-MSFT commented:
Gentle ping for this week, thank you

vvolam commented:
@nissampa could you please rebase the PR as discussed over mail and also please address any outstanding comments. Thank you!

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

nissampa commented:
> @nissampa could you please rebase the PR as discussed over mail and also please address any outstanding comments. Thank you!

Addressed the remaining comments that we talked in the call and yes rebased the PR as well.

JibinBao commented on tests/smartswitch/common/device_utils_dpu.py:
```diff
@@ -128,39 +136,232 @@ def check_dpu_module_status(duthost, power_status, dpu_name):
+        Returns True or False based on pmon status
+    """
+
+    output_pmon_status = duthost.command('docker ps')
```
maybe you can simply your code as follows:
 output_pmon_status = duthost.command('docker ps | grep pmon' )['stdout']
and then check  if "Up" in output_pmon_status

nissampa replied:
Changed it.

JibinBao commented on tests/smartswitch/platform_tests/test_reload_dpu.py:
```diff
@@ -4,71 +4,323 @@
+                                     platform_api_conn, num_dpu_modules):  # noqa: F811, E501
+    """
+    @summary: Test memory exhaustion on NPU by running a heavy process,
+              followed by a reboot of the NPU.
```
NPU reboot caused by running a heavy process?

nissampa replied:
Yes correct. Changed the description appropriately.

JibinBao commented on tests/smartswitch/platform_tests/test_reload_dpu.py:
```diff
@@ -4,71 +4,323 @@
+                                platform_api_conn, num_dpu_modules):  # noqa: F811, E501
+    """
+    @summary: Test NPU recovery from a kernel panic,
+              followed by a reboot of the NPU.
```
it means that a  kernel panic will cause NPU reboot

nissampa replied:
Changed the description.

nissampa commented on tests/smartswitch/platform_tests/test_reload_dpu.py:
```diff
@@ -4,71 +4,323 @@
+                               localhost,
+                               platform_api_conn, num_dpu_modules):  # noqa F811, E501
     """
     @summary: Verify output of `config chassis modules startup <DPU_Number>`
```
Changed it.

mssonicbld commented:
/azp run

nissampa commented on tests/smartswitch/common/device_utils_dpu.py:
```diff
@@ -128,39 +136,232 @@ def check_dpu_module_status(duthost, power_status, dpu_name):
+    return False
+
+
+def execute_dpu_commands(duthost, ipaddress, command, output=True):
```
Can I come back to it after this commit ?
Add this feature on case by case basis and for future cases ?

Since only two cases are involved in executing directly on DPU and that also involves executing commands on both NPU and DPU in the same cases.

JibinBao replied:
OK. It is your call.

mssonicbld commented:
/azp run

vvolam commented on tests/smartswitch/platform_tests/test_reload_dpu.py:
```diff
@@ -4,71 +4,324 @@
+                              dpu_off_list, ip_address_list)
+
+
+def test_kernel_panic_on_switch(duthosts,
```
There is some code duplication from test_memory_exhaustion_on_switch(), could your define new functions for these common codes?

nissampa replied:
Modified it

vvolam commented on tests/smartswitch/platform_tests/test_reload_dpu.py:
```diff
@@ -4,71 +4,324 @@
+    pytest_assert(ping_status == 1, "Ping to one or more DPUs has failed")
+
+
+def test_memory_exhaustion_on_dpu(duthosts, enum_rand_one_per_hwsku_hostname,
```
Same here, code duplication with test_kernel_panic_on_dpu(). could you define functions for common code?

nissampa replied:
Modified it.

vvolam reviewed (commented):
On Mellanox testbed, new tests in test_reload_dpu.py tests are failing for memory exhaustion and kernel panic as the following command is failing after reload. 

"docker exec -i pmon supervisorctl reread"

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

JibinBao commented on tests/smartswitch/common/device_utils_dpu.py:
```diff
@@ -128,39 +136,232 @@ def check_dpu_module_status(duthost, power_status, dpu_name):
+    Returns:
+        Returns the output of the given command
+    """
+    username = duthost.facts['ssh_dpu']['username']
```
When can passwordless ssh be ready?
Before passwordless ssh is ready, how to run the test?

nissampa replied:
@vvolam Vasundhara is in talks with Price George and Lin regarding the design of ssh usage.

If it is recommended to use dpu fixtures, I will change it accordingly.

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

nissampa commented:
@vvolam @JibinBao 

I removed the execute_dpu_commands and used dpuhosts fixtures for running commands on DPU.

Let me know if things are good and please give an approval.

JibinBao commented:
> @vvolam @JibinBao
> 
> I removed the execute_dpu_commands and used dpuhosts fixtures for running commands on DPU.
> 
> Let me know if things are good and please give an approval.

Hi @nissampa , After you use dpuhost fixutres, did you test it, if test pass, I think it is ok.

nissampa commented:
> > @vvolam @JibinBao
> > I removed the execute_dpu_commands and used dpuhosts fixtures for running commands on DPU.
> > Let me know if things are good and please give an approval.
> 
> Hi @nissampa , After you use dpuhost fixutres, did you test it, if test pass, I think it is ok.

Yes I tested it and tests are passing with dpuhost fixtures.

JibinBao commented:
> > > @vvolam @JibinBao
> > > I removed the execute_dpu_commands and used dpuhosts fixtures for running commands on DPU.
> > > Let me know if things are good and please give an approval.
> > 
> > 
> > Hi @nissampa , After you use dpuhost fixutres, did you test it, if test pass, I think it is ok.
> 
> Yes I tested it and tests are passing with dpuhost fixtures.

Thanks.

mssonicbld commented:
/azp run

vvolam commented:
@JibinBao Are these tests verified on NVIDIA smartswitch? I am running into failure when running "config reload -y" command on our switch.

Here are the logs:

>           raise RunAnsibleModuleFail("run module {} failed".format(self.module_name), res)
E           tests.common.errors.RunAnsibleModuleFail: run module shell failed, Ansible Results =>
E           failed = True
E           changed = True
E           rc = 1
E           cmd = sudo config reload -y
E           start = 2025-02-26 00:21:41.300903
E           end = 2025-02-26 00:22:05.854479
E           delta = 0:00:24.553576
E           msg = non-zero return code
E           invocation = {'module_args': {'executable': '/bin/bash', '_raw_params': 'sudo config reload -y', '_uses_shell': True, 'warn': False, 'stdin_add_newline': True, 'strip_empty_ends': True, 'argv': None, 'chdir': None, 'creates': None, 'removes': None, 'stdin': None}}
E           _ansible_no_log = None
E           stdout =
E           Acquired lock on /etc/sonic/reload.lock
E           Disabling container and routeCheck monitoring ...
E           Stopping SONiC target ...
E           Running command: /usr/local/bin/sonic-cfggen -j /etc/sonic/init_cfg.json -j /etc/sonic/config_db.json --write-to-db
E           Running command: /usr/local/bin/db_migrator.py -o migrate
E           sonic_yang(3):Data Loading Failed:Leafref "/sonic-dhcp-server-ipv4:sonic-dhcp-server-ipv4/sonic-dhcp-server-ipv4:DHCP_SERVER_IPV4_CUSTOMIZED_OPTIONS/sonic-dhcp-server-ipv4:DHCP_SERVER_IPV4_CUSTOMIZED_OPTIONS_LIST/sonic-dhcp-server-ipv4:name" of value "option223" points to a non-existing leaf.
E           Yang validation failed


I suspect this is due to a recent sonic-buildimage change but want to confirm if you are hitting same issue.

nissampa commented:
@vvolam, Can you help me in approving this set if you feel that code wise things are good ?

vvolam commented:
> @vvolam, Can you help me in approving this set if you feel that code wise things are good ?

@nissampa After fixing config reload issue manually, I now observe times are not sufficient on NVIDIA switch. 

@JibinBao  could you please test and confirm the times required?

vvolam commented on tests/smartswitch/platform_tests/test_show_platform_dpu.py:
```diff
@@ -72,18 +80,23 @@ def test_reboot_cause(duthosts, enum_rand_one_per_hwsku_hostname,
-                                 check_dpu_module_status,  # noqa: F405
+        pytest_assert(wait_until(DPU_MAX_TIMEOUT, DPU_TIME_INT, 0,
+                                 check_dpu_module_status,
                                  duthost, "off",
```
Is "off" here Admin status?? If so, it should be "down"!

nissampa replied:
It is not admin status here. It represents the operation we do in the above step and sending it as argument to check for the dpu status reflecting the expected status with respect to that operation.

JibinBao commented:
> > @vvolam, Can you help me in approving this set if you feel that code wise things are good ?
> 
> @nissampa After fixing config reload issue manually, I now observe times are not sufficient on NVIDIA switch.
> 
> @JibinBao could you please test and confirm the times required?

Next week, I will try running the tests on NVIDIA Switch

vvolam commented on tests/smartswitch/platform_tests/test_show_platform_dpu.py:
```diff
@@ -72,18 +80,23 @@ def test_reboot_cause(duthosts, enum_rand_one_per_hwsku_hostname,
+        duthost.shell("config chassis modules startup %s" % (dpu_name))
+        pytest_assert(wait_until(DPU_MAX_TIMEOUT, DPU_TIME_INT, 0,
+                      check_dpu_module_status,
+                      duthost, "on", dpu_name),
```
What is "on" here? If Admin-Status, it will be displayed as "up"

nissampa replied:
Same comment as above.

mssonicbld commented:
/azp run

nissampa commented on tests/smartswitch/common/device_utils_dpu.py:
```diff
@@ -128,39 +150,312 @@ def check_dpu_module_status(duthost, power_status, dpu_name):
             'show chassis module status | grep %s' % (dpu_name))
 
     if "Offline" in output_dpu_status["stdout"]:
```
Changed it to check lower cases

nissampa commented on tests/smartswitch/platform_tests/test_show_platform_dpu.py:
```diff
@@ -100,30 +113,227 @@ def test_pcie_link(duthosts, enum_rand_one_per_hwsku_hostname,
+            continue
+        ip_address = module.get_midplane_ip(platform_api_conn, index)
+
+        logging.info("Checking show system-health summary on DPU")
```
Added the dpu names to the logs

mssonicbld commented:
/azp run

nissampa commented:
/azpw run

mssonicbld commented:
/AzurePipelines run

Nishanths1992 commented:
/azpw run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

Nishanths1992 commented:
/azpw run

Nishanths1992 commented:
/azpw run

JibinBao commented on tests/smartswitch/platform_tests/test_reload_dpu.py:
```diff
@@ -4,71 +4,182 @@
 from tests.common.platform.device_utils import platform_api_conn  # noqa: F401,F403
 
 pytestmark = [
     pytest.mark.topology('smartswitch')
```
Can you add  t1 here, because we run the test on t1-lag-28.  
pytest.mark.topology('smartswitch', 't1')
or 
pytest.mark.topology('smartswitch', 'any')
Actually the test is not related to topo, it shoud be top any,  but it is related to platform

nissampa replied:
@vvolam, Can I add 'any' in here ?

vvolam replied:
@JibinBao I think we wanted to run these tests only on smartswitch and it is t1 topo as well.

If "any" is added, tests will try to run on all platforms. Please correct me if I am missing something

JibinBao replied:
Hi @vvolam, You mean when pytest.mark.topology('smartswitch'), the case also can run on t1 topo. Is that correct? 
Can you confirm it?
But in our local setup, I ran it on the t1-lag-28 topo, the case was skipped.  
'any' is becuase the case is not related to topo, it is just related to platform.

JibinBao commented on tests/smartswitch/platform_tests/test_reload_dpu.py:
```diff
@@ -4,71 +4,182 @@
+    for index in range(len(dpu_on_list)):
+        logging.info("Triggering Kernel Panic on %s" % (dpu_on_list[index]))
+        ip_address = ip_address_list[index]
+        dpu_number = int(ip_address[-1])
```
get dpu number by ip_address[-1]  is not accurate
For example, for Nvidia smartswitch, we have 4 dpus. 
The first dpu ip  is '169.254.200.1', and the fourth dpu ip is  '169.254.200.4'. As the logic, dpu_nubmer will 1, will use dpuhosts[1]  represent dpu0, but it might be dpu1. And when dpu_number  is 4, it will raise the error "IndexError: list index out of range", because the dpuhosts only have 4 dpus

Nishanths1992 replied:
Changed it to get it from dpu name itself.

JibinBao reviewed (commented):
After running the test, and find new issue. Please update it. Thanks

Nishanths1992 commented on tests/smartswitch/platform_tests/test_reload_dpu.py:
```diff
@@ -4,71 +4,182 @@
+                "Triggering Memory Exhaustion on %s" % (dpu_on_list[index])
+                )
+        ip_address = ip_address_list[index]
+        dpu_number = int(ip_address[-1])
```
Changed it to get it from dpu name itself.

mssonicbld commented:
/azp run

vvolam reviewed (commented):
Mostly cosmetic comments.

nissampa commented on tests/smartswitch/common/device_utils_dpu.py:
```diff
@@ -127,40 +149,313 @@ def check_dpu_module_status(duthost, power_status, dpu_name):
+                      "DPU is not operationally down")
+
+    ping_status = check_dpu_ping_status(duthost, ip_address_list)
+    pytest_assert(ping_status == 1, "Ping to DPU has failed")
```
added log failure in dpu ping function itself.

nissampa commented on tests/smartswitch/common/device_utils_dpu.py:
```diff
@@ -127,40 +149,313 @@ def check_dpu_module_status(duthost, power_status, dpu_name):
+
+        logging.info("Shutting down %s" % (dpu_on_list[index]))
+        duthost.shell(
+                "config chassis modules shutdown %s" % (dpu_on_list[index])
```
I guess we don't need it. In my tests, it ran without sudo itself.

nissampa commented on tests/smartswitch/common/device_utils_dpu.py:
```diff
@@ -127,40 +149,313 @@ def check_dpu_module_status(duthost, power_status, dpu_name):
+       Returns Nothing
+    """
+    logging.info(f"Checking system-health status of {dpu_name}")
+    output_dpu_health_status = duthost.show_and_parse(f"show system-health dpu {dpu_name}")
```
Same as other reply.

nissampa commented on tests/smartswitch/platform_tests/test_reload_dpu.py:
```diff
@@ -4,71 +4,182 @@
 ]
 
+kernel_panic_cmd = "nohup bash -c 'sleep 5 && echo c > /proc/sysrq-trigger' &"
+memory_exhaustion_cmd = "nohup bash -c 'sleep 5 && tail /dev/zero' &"
```
Same reply for other sudo comments

vvolam replied:
@nissampa We use sudo to make sure we don't run into privilege issues. Could you add them?

nissampa replied:
added it.

nissampa commented on tests/smartswitch/common/device_utils_dpu.py:
```diff
@@ -127,40 +149,313 @@ def check_dpu_module_status(duthost, power_status, dpu_name):
+                        dpu_on_list, dpu_off_list,
+                        ip_address_list):
+    """
+    Checks DPU off/on and reboot cause status Post Test
```
Yes. I updated the status.

mssonicbld commented:
/azp run

nissampa commented on tests/smartswitch/platform_tests/test_reload_dpu.py:
```diff
@@ -4,71 +4,182 @@
+                           ip_address_list)
+
+
+def test_kernel_panic_on_switch(duthosts,
```
Changed all the test cases in the file to make it more uniform.

nissampa commented on tests/smartswitch/platform_tests/test_reload_dpu.py:
```diff
@@ -4,71 +4,182 @@
+                              dpu_off_list, ip_address_list)
+
+
+def test_memory_exhaustion_on_switch(duthosts,
```
Changed all the test cases in the file to make it more uniform.

nissampa commented on tests/smartswitch/platform_tests/test_reload_dpu.py:
```diff
@@ -4,71 +4,182 @@
+memory_exhaustion_cmd = "nohup bash -c 'sleep 5 && tail /dev/zero' &"
+
 
 def test_dpu_ping_after_reboot(duthosts, enum_rand_one_per_hwsku_hostname,
```
Changed all the test cases in the file to make it more uniform.

mssonicbld commented:
/azp run

nissampa commented on tests/run_tests.sh:
```diff
@@ -13,6 +13,7 @@ function show_help_and_exit()
     echo "    -e <parameters>: specify extra parameter(s) (default: none)"
```
e is already been in use right.

JibinBao commented on tests/smartswitch/platform_tests/test_show_platform_dpu.py:
```diff
@@ -4,17 +4,27 @@
-from tests.common.devices.sonic import *  # noqa: 403
 
 pytestmark = [
     pytest.mark.topology('smartswitch')
```
same question as that in test_reload_dpu.py.  Please also upate it

nissampa replied:
added it.

nissampa replied:
added it.

nissampa commented:
/azpw run

mssonicbld commented:
/AzurePipelines run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

nissampa commented:
/azpw run

mssonicbld commented:
/AzurePipelines run

nissampa commented:
@vvolam, Can you approve it if things are good ?

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

nissampa commented:
/azpw run

mssonicbld commented:
/AzurePipelines run

vvolam commented on tests/smartswitch/common/device_utils_dpu.py:
```diff
@@ -127,40 +151,314 @@ def check_dpu_module_status(duthost, power_status, dpu_name):
+        pytest_assert(wait_until(DPU_MAX_TIMEOUT, DPU_MAX_TIME_INT, 0,
+                      check_dpu_module_status,
+                      duthost, "off", dpu_on_list[index]),
+                      "DPU is not down after memory exhaustion")
```
This code is not common to both memory exhaustion and kernel  panic. could you fix the log accordingly?

nissampa replied:
Changed it.

vvolam commented on tests/smartswitch/common/device_utils_dpu.py:
```diff
@@ -127,40 +151,314 @@ def check_dpu_module_status(duthost, power_status, dpu_name):
+                      "DPU is not down after memory exhaustion")
+
+        logging.info("Shutting down %s" % (dpu_on_list[index]))
+        duthost.shell(
```
Why do we want to manually shut down? Don't the DPU go down after memory exhaustion or kernel panic, we are already asserting above if not.

nissampa replied:
That is because, pcie link would be down after kernel panic or memory exhaustion.

vvolam replied:
Can we discuss this function in tomorrow's call for a better understanding?

vvolam reviewed (commented):
I am a bit confused on steps in post_test_dpu_check()

Should this be just waiting for DPU to power on after a -ve test case or a reboot and then just the ping? If not, fail the test?

gpunathilell commented:
In test_show_platform_dpu.py is it possible to execute the config startup and shutdown commands for the DPUs in parallel?
Doing it sequentially increases the test time by a lot (The worst case time taken is 
DPU_MAX_TIMEOUT * Number of DPUs * 2 (power off+ power on)  * 3 (3 tests which run power off + Power off) for all tests in this file. Executing it parallely for each DPU makes the most sense
Also we can reduce the value for `DPU_TIME_INT` to 20 or 30, As the interval to check can be much smaller,

mssonicbld commented:
/azp run

nissampa commented on tests/smartswitch/platform_tests/test_show_platform_dpu.py:
```diff
@@ -4,17 +4,27 @@
 
```
Sure, changed it.

mssonicbld commented:
/azp run

nissampa commented:
> I am a bit confused on steps in post_test_dpu_check()
> 
> Should this be just waiting for DPU to power on after a -ve test case or a reboot and then just the ping? If not, fail the test?

That post check used on after couple of operations like men exhaustion and on kernel panic.
Both the cases, after the operation DPU goes for a reboot and comes up with pcie link down. Hence, checking for offline status.
Upon shutdown and power up of DPU, pcie link gets restored back.

nissampa commented:
@vvolam Addressed the comments and updated the PR. 
Please let me know if it is good for approval.

mssonicbld commented:
/azp run

vvolam commented on tests/smartswitch/platform_tests/test_platform_dpu.py:
```diff
@@ -0,0 +1,352 @@
+        pytest_assert(wait_until(DPU_MAX_TIMEOUT, DPU_TIME_INT, 0,
+                      check_dpu_module_status,
+                      duthost, "on", dpu_name),
+                      "DPU is not operationally up")
```
could you add DPU ping status and critical_process check as well here?

gpunathilell commented on tests/smartswitch/platform_tests/test_platform_dpu.py:
```diff
@@ -0,0 +1,352 @@
+        duthost.shell("sudo config chassis modules shutdown %s" % (dpu_name))
+
+    for index in range(num_dpu_modules):
+        pytest_assert(wait_until(DPU_MAX_TIMEOUT, DPU_TIME_INT, 0,
```
We need separate threads, this way assuming first DPU takes `DPU_MAX_TIMEOUT` to shutdown, the we wait for the second DPU from that point so essentially the timeout is `DPU_MAX_TIMEOUT*2` which would be wrong.

nissampa replied:
Removed this case altogether since it is getting covered as part of many other cases.

gpunathilell replied:
This affects all test cases in the file, since we are using the `DPU_MAX_TIMEOUT` to monitor the DPU will power on/off within the specified time, we need multi-thread implementation for atleast one of the test cases so that all DPUs are monitored for maximum boot time/shutdown time (since it would be a better design, it can be taken over for all the test cases as well)

mssonicbld commented:
/azp run

nissampa commented on tests/smartswitch/common/device_utils_dpu.py:
```diff
@@ -127,40 +151,321 @@ def check_dpu_module_status(duthost, power_status, dpu_name):
+    """
+    Checks DPU off/on and reboot cause status Post Test
+    Args:
+       duthost: Host handle
```
Will change it.

nissampa commented on tests/smartswitch/common/device_utils_dpu.py:
```diff
@@ -3,11 +3,31 @@
+INTF_MAX_TIMEOUT = 300
+INTF_TIME_INT = 5
+DPU_MAX_TIMEOUT = 360
+DPU_MAX_TIME_INT = 120
```
will change it.

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

JibinBao commented on tests/smartswitch/platform_tests/test_show_platform_dpu.py:
```diff
@@ -117,13 +122,224 @@ def test_pcie_link(duthosts, enum_rand_one_per_hwsku_hostname,
+        if rc:
+            continue
+
+        command = ('python -c "import pexpect; '
```
Can you  add 'sudo' before 'python'?
I tested it on our device,  it will raise exception if there is not sudo
'sudo python -c  ...'

nissampa replied:
added it.

JibinBao commented on tests/smartswitch/common/device_utils_dpu.py:
```diff
@@ -127,40 +151,322 @@ def check_dpu_module_status(duthost, power_status, dpu_name):
+        dpu_number = int(re.search(r'\d+', dpu_on_list[index]).group())
+        logging.info("Checking crictical processes \
+                      on %s" % (dpu_on_list[index]))
+        check_dpu_critical_processes(dpuhosts, dpu_number)
```
It is better to use wait_until, because we cannot guarantee that all processed are ready at that time

nissampa replied:
Added it.

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

nissampa commented:
/azpw run

mssonicbld commented:
/AzurePipelines run

mssonicbld commented:
/azp run

nissampa commented:
/azpw run

mssonicbld commented:
/AzurePipelines run

nissampa commented:
/azpw run

mssonicbld commented:
/AzurePipelines run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

nissampa commented:
/azpw run

vvolam commented:
@nissampa could you please merge latest master to fix this build issue?

mssonicbld commented:
/azp run

nissampa commented:
> @nissampa could you please merge latest master to fix this build issue?

@vvolam Merged the latest master, Still I see failures.

mssonicbld commented:
/AzurePipelines run

JibinBao commented:
Hi @nissampa 
Some case has issue, I have a fix,  can you review it: https://github.com/sonic-net/sonic-mgmt/pull/18185/?
